src/raftstore/store/util.rs

```diff
@@ -177,22 +181,188 @@ pub fn get_region_approximate_size(db: &DB, region: &metapb::Region) -> Result<u
     Ok(size)
 }
 
+/// Lease records an expired time, for examining the current moment is in lease or not.
+/// It's dedicated to the Raft leader lease mechanism, contains either state of
+///   1. Suspect Timestamp
+///      A suspicious leader lease timestamp, which marks the leader may still hold or lose
+///      its lease until the clock time goes over this timestamp.
+///   2. Valid Timestamp
+///      A valid leader lease timestamp, which marks the leader holds the lease for now.
+///      The lease is valid until the clock time goes over this timestamp.
+///
+/// ```text
+/// Time
+/// |---------------------------------->
+///         ^               ^
+///        Now           Suspect TS
+/// State:  |    Suspect    |   Suspect
+///
+/// |---------------------------------->
+///         ^               ^
+///        Now           Valid TS
+/// State:  |     Valid     |   Expired
+/// ```
+///
+/// Note:
+///   - Valid timestamp would increase when raft log entries are applied in current term.
+///   - Suspect timestamp would be set after the message `MsgTimeoutNow` is sent by current peer.
+///     The message `MsgTimeoutNow` starts a leader transfer procedure. During this procedure,
+///     current peer as an old leader may still hold its lease or lose it.
+///     It's possible there is a new leader elected and current peer as an old leader
+///     doesn't step down due to network partition from the new leader. In that case,
+///     current peer lose its leader lease.
+///     Within this suspect leader lease expire time, read requests could not be performed
+///     locally.
+///   - The valid leader lease should be `lease = max_lease - (quorum_commit - send_to_quorum)`
+///     And the expired timestamp for that leader lease is `quorum_commit + lease`,
+///     which is `send_to_quorum + max_lease` in short.
+// TODO: add a remote Lease. A special lease that derives from Lease, it will be sent
+//       to the local read thread, so name it remote. If Lease expires, the remote must
+//       expire too.
+pub struct Lease {
+    // A suspect timestamp is in the Either::Left(_),
+    // a valid timestamp is in the Either::Right(_).
+    bound: Option<Either<Timespec, Timespec>>,
+    max_lease: Duration,
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub enum LeaseState {
+    /// The lease is suspicious, may be invalid.
+    Suspect,
+    /// The lease is valid.
+    Valid,
+    /// The lease is expired.
+    Expired,
+}
+
+impl Lease {
+    pub fn new(max_lease: Duration) -> Lease {
+        Lease {
+            bound: None,
+            max_lease: max_lease,
+        }
+    }
+
+    fn next_expired_time(&self, send_to_quorum: Timespec) -> Timespec {
+        send_to_quorum + self.max_lease
+    }
+
+    /// Renew the lease to the `send_to_quorum + max_lease`.
+    pub fn renew(&mut self, send_to_quorum: Timespec) {
```

Comment:
What does it mean by `send_to_quorum`? Is it the time that leader send out the proposal or the time when quorum receive it?